### PR TITLE
[GSoC 2022] nasm/simd support for libjpeg-turbo

### PR DESCRIPTION
**Pull request by ocpalo**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

Related with #22344

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -137,6 +137,15 @@ if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)  # https://cmake.org/cmake/help/
   endif()
 endif()
 
+if(BUILD_JPEG AND WITH_LIBJEG_TURBO_SIMD)
```
Comment (ocpalo) [inline CMakeLists.txt:140]: Need this to set right status for SIMD Support in the line 1353

Comment (sturkmen72) [reply CMakeLists.txt:140]: **LGTM** ( i tested only on Windows )

you can see build logs with [nasm is available](https://ci.appveyor.com/project/sturkmen72/opencv-image-reading/builds/44454473?fullLog=true) and [nasm is unavailable](https://ci.appveyor.com/project/sturkmen72/opencv-image-reading/builds/44454275?fullLog=true)

result output of [test code](https://github.com/sturkmen72/opencv-image-reading/blob/jpg/test-2.cpp) is like below

**SIMD Support:            YES**
```
Frames captured (VideoCapture) :   100    Average FPS:     170.8    Average time per frame:      5.85 ms
Frames captured (imread)       :   100    Average FPS:     109.0    Average time per frame:      9.17 ms
```

**SIMD Support:            NO**
```
Frames captured (VideoCapture) :   100    Average FPS:     171.1    Average time per frame:      5.84 ms
Frames captured (imread)       :   100    Average FPS:      82.9    Average time per frame:     12.06 ms
```

**Comment by sturkmen72** [comment]
**LGTM** ( i tested only on Windows )

you can see build logs with [nasm is available](https://ci.appveyor.com/project/sturkmen72/opencv-image-reading/builds/44454473?fullLog=true) and [nasm is unavailable](https://ci.appveyor.com/project/sturkmen72/opencv-image-reading/builds/44454275?fullLog=true)

result output of [test code](https://github.com/sturkmen72/opencv-image-reading/blob/jpg/test-2.cpp) is like below

**SIMD Support:            YES**
```
Frames captured (VideoCapture) :   100    Average FPS:     170.8    Average time per frame:      5.85 ms
Frames captured (imread)       :   100    Average FPS:     109.0    Average time per frame:      9.17 ms
```

**SIMD Support:            NO**
```
Frames captured (VideoCapture) :   100    Average FPS:     171.1    Average time per frame:      5.84 ms
Frames captured (imread)       :   100    Average FPS:      82.9    Average time per frame:     12.06 ms
```

File: modules/imgcodecs/perf/perf_jpeg.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+    using namespace perf;
+
+    typedef perf::TestBaseWithParam<std::string> JPEG;
+
+    PERF_TEST(JPEG, Decode)
+    {
+        String filename = getDataPath("stitching/boat1.jpg");
+
+        TEST_CYCLE() imread(filename);
```
Comment (asmorkalov) [inline modules/imgcodecs/perf/perf_jpeg.cpp:16]: It makes sense to benchmark imdecode and imencode. They do not read from /write to disk and benchmarks shows pure time.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by ocpalo** [comment]
* Without SIMD:
Screenshot from 2022-08-15 18-53-53


* With SIMD
Screenshot from 2022-08-15 18-51-15

**Comment by asmorkalov** [comment]
FYI: perf tests can report status and performance statistics with to xml with flag --gtest_output=xml:report_file_name.xml. The xml reports could be compared with summary.py script: https://github.com/opencv/opencv/blob/4.x/modules/ts/misc/summary.py

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 3]
```diff
@@ -15,8 +15,68 @@ endif()
...
+
+message(STATUS "${BITS}-bit build (${CPU_TYPE})")
+
+macro(boolean_number var)
+  if(${var})
+    set(${var} 1 ${ARGN})
+  else()
+    set(${var} 0 ${ARGN})
+  endif()
+endmacro()
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:74]: The macro is not used in the patch.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 4]
```diff
@@ -15,8 +15,68 @@ endif()
...
+        CMAKE_OSX_ARCHITECTURES MATCHES "arm64" OR
+        CMAKE_OSX_ARCHITECTURES MATCHES "i386")
+  set(CPU_TYPE ${CMAKE_OSX_ARCHITECTURES})
+endif()
+if(CMAKE_OSX_ARCHITECTURES MATCHES "ppc")
+  set(CPU_TYPE powerpc)
+endif()
+if(MSVC_IDE AND CMAKE_GENERATOR_PLATFORM MATCHES "arm64")
+  set(CPU_TYPE arm64)
+endif()
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:64]: OpenCV already has processor options detector in CMake. Please reuse variables like X86_64, ARM, PPC64 and others from https://github.com/opencv/opencv/blob/4.x/cmake/OpenCVDetectCXXCompiler.cmake.

Comment (asmorkalov) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt:64]: Pay attention that NEON, AVX512, etc support are managed with dedicated flags.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt (hunk 3 above)
Comment (ocpalo) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt:74]: It is actually used by simd/CMakeLists.txt [here](https://github.com/opencv/opencv/blob/6aa3e9ff63fcf5e584d9366543005ed07cdc9908/3rdparty/libjpeg-turbo/src/simd/CMakeLists.txt#L363) . I did not want to modify that file. That is why I included it inside that CMakeList.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt (hunk 4 above)
Comment (ocpalo) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt]: > Pay attention that NEON, AVX512, etc support are managed with dedicated flags.

Could you eloborate more this? I did not get it what you mean by that.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt (hunk 3 above)
Comment (ocpalo) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt]: I carried this into simd/CMakeList

**Comment by ocpalo** [comment]
I have changed some parts of the CMakeLists take into account your comments. But i am thinking that is not what you want exactly. Can you give me more info about that and if possible, could you look at last workflow errors? In the macOS ARM build, 3 tests are failed. I did not understand why this is the case because these pass in my computer. @asmorkalov

File: 3rdparty/libjpeg-turbo/CMakeLists.txt (hunk 4 above)
Comment (asmorkalov) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt]: https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options

Comment (asmorkalov) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt]: Briefly, NEON,  AVX2, AVX512, etc support are not guaranteed on all CPUs. There are Atom chips, even without AVX2 support. The same for NEON. Some NVIDIA Tegra chips has arm-v7a architecture, but does not implement NEON extension. NEON code will raise SIGILL here.
`CPU_BASELINE_FINAL` contains all hardware specific optimizations that are enabled by default and does not participate in dynamic dispatch. libjpeg turbo should use the list of options to enable proper optimizations.

Comment (asmorkalov) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt]: For example you can take a look on Carotene HAL management for ARM: https://github.com/opencv/opencv/blob/9fcf0152140c449f74678add820e81de3e69b8b0/CMakeLists.txt#L880

Comment (ocpalo) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt]: If I understand your concern correctly, NEON and others are handled by libjpeg-turbo simd/cmakelists. 

For example, [here](https://github.com/opencv/opencv/blob/79afd80817c8d84484052ae06e64d61db202d27a/3rdparty/libjpeg-turbo/src/simd/CMakeLists.txt#L218) NEON support are handled below. 

I will investigave further. Thanks for the link.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 5]
```diff
@@ -15,8 +15,54 @@ endif()
...
+  math(EXPR COUNT "${COUNT}+1")
+endforeach()
+
+if(X86_64 OR X86 OR CMAKE_SYSTEM_PROCESSOR_LC MATCHES "ia32")
+  if(BITS EQUAL 64 OR CMAKE_C_COMPILER_ABI MATCHES "ELF X32")
+    set(CPU_TYPE x86_64)
+  else()
+    set(CPU_TYPE i386)
+  endif()
+  if(NOT CMAKE_SYSTEM_PROCESSOR STREQUAL ${CPU_TYPE})
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:34]: Looks strange. Why do you need this?

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 6]
```diff
@@ -15,8 +15,54 @@ endif()
...
 
+math(EXPR BITS "${CMAKE_SIZEOF_VOID_P} * 8")
+string(TOLOWER ${CMAKE_SYSTEM_PROCESSOR} CMAKE_SYSTEM_PROCESSOR_LC)
+set(COUNT 1)
+foreach(ARCH ${CMAKE_OSX_ARCHITECTURES})
+  if(COUNT GREATER 1)
+    message(FATAL_ERROR "The libjpeg-turbo build system does not support multiple values in CMAKE_OSX_ARCHITECTURES.")
+  endif()
+  math(EXPR COUNT "${COUNT}+1")
+endforeach()
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:26]: OpenCV CMake system does not support build for several architectures in a single step. I do not think that the check is required.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 7]
```diff
@@ -15,8 +15,54 @@ endif()
...
+
+if(X86_64 OR X86 OR CMAKE_SYSTEM_PROCESSOR_LC MATCHES "ia32")
+  if(BITS EQUAL 64 OR CMAKE_C_COMPILER_ABI MATCHES "ELF X32")
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:29]: `CMAKE_SIZEOF_VOID_P` is already taken in account when X86_64 and X86 are set. You do not need to warry about it here. See https://github.com/opencv/opencv/blob/9fcf0152140c449f74678add820e81de3e69b8b0/cmake/OpenCVDetectCXXCompiler.cmake#L121

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 8]
```diff
@@ -15,8 +15,54 @@ endif()
...
+    set(CPU_TYPE x86_64)
+  else()
+    set(CPU_TYPE i386)
+  endif()
+  if(NOT CMAKE_SYSTEM_PROCESSOR STREQUAL ${CPU_TYPE})
+    set(CMAKE_SYSTEM_PROCESSOR ${CPU_TYPE})
+  endif()
+elseif(AARCH64 OR ARM)
+  if(BITS EQUAL 64)
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:38]: The same for ARM as for x86. See https://github.com/opencv/opencv/blob/9fcf0152140c449f74678add820e81de3e69b8b0/cmake/OpenCVDetectCXXCompiler.cmake#L121

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 9]
```diff
@@ -15,8 +15,54 @@ endif()
...
+  set(CPU_TYPE powerpc)
+else()
+  set(CPU_TYPE ${CMAKE_SYSTEM_PROCESSOR_LC})
+endif()
+
+if(CMAKE_OSX_ARCHITECTURES MATCHES "x86_64" OR
+        CMAKE_OSX_ARCHITECTURES MATCHES "arm64" OR
+        CMAKE_OSX_ARCHITECTURES MATCHES "i386")
+  set(CPU_TYPE ${CMAKE_OSX_ARCHITECTURES})
+endif()
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:53]: X86, X86_64, AARCH64, etc flags are set for Apple systems too. You do not need to handle it separately.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 10]
```diff
@@ -15,8 +15,54 @@ endif()
...
+if(CMAKE_OSX_ARCHITECTURES MATCHES "ppc")
+  set(CPU_TYPE powerpc)
+endif()
+if(MSVC_IDE AND CMAKE_GENERATOR_PLATFORM MATCHES "arm64")
+  set(CPU_TYPE arm64)
+endif()
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:59]: On Windows as well.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 11]
```diff
@@ -15,8 +15,54 @@ endif()
...
+
+message(STATUS "${BITS}-bit build (${CPU_TYPE})")
+
 option(WITH_ARITH_ENC "Include arithmetic encoding support when emulating the libjpeg v6b API/ABI" TRUE)
 option(WITH_ARITH_DEC "Include arithmetic decoding support when emulating the libjpeg v6b API/ABI" TRUE)
+option(WITH_LIBJPEG_TURBO_SIMD "Include SIMD extensions, if available for this platform" TRUE)
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:65]: OpenCV has `CV_DISABLE_OPTIMIZATION` flag in CMake to force disable platform specific optimizations. It's used for debugging purpose mostly, to catch SIMD path misbehavior and compare performance. I propose to replace default value by `(NOT CV_DISABLE_OPTIMIZATION)` how it's done for Carotete: https://github.com/opencv/opencv/blob/9fcf0152140c449f74678add820e81de3e69b8b0/CMakeLists.txt#L246

Comment (asmorkalov) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt:65]: Also optimization baseline should be checked.

**Comment by sturkmen72** [comment]
image
I guess the simd related codes do not appear in the relevant place in Solution Explorer. something needs to be done to solve it

**Comment by ocpalo** [comment]
Geometric mean (ms)

|Name of Test|libjpeg|libjpeg-turbo|libjpeg-turbo vs previous (x-factor)|
|---|:-:|:-:|:-:|
|Decode::JPEG|62.426 ms|38.651 ms|1.62|
|Encode::JPEG|58.559 ms|33.405 ms|1.75|

**Comment by ocpalo** [comment]
I think libjpeg-turbo with SIMD is broken if using with MJPEG backend.

**Comment by ocpalo** [comment]
For example, building for Windows Simd option on, 

```cpp
[ RUN      ] videoio/videoio_bunny.read_position/31, where GetParam() = ("mjpg.avi", CV_MJPEG)
Command exited with code -1073740940
```

This test fails. 

And in macOS, i386 arch,

```cpp
[ RUN      ] videoio/videoio_synthetic.write_read_position/39, where GetParam() = (976x768, FOURCC(MJPG), .avi, CV_MJPEG, 30dB)
/Users/appveyor/projects/opencv-image-reading/opencv/modules/videoio/test/test_video_io.cpp:289: Failure
Expected: (psnr) > (PSNR_GT), actual: 27.7196 vs 30
frame 0
Google Test trace:
/Users/appveyor/projects/opencv-image-reading/opencv/modules/videoio/test/test_video_io.cpp:72: consecutive read
/Users/appveyor/projects/opencv-image-reading/opencv/modules/videoio/test/test_video_io.cpp:289: Failure
Expected: (psnr) > (PSNR_GT), actual: 27.7196 vs 30
frame 0
Google Test trace:
/Users/appveyor/projects/opencv-image-reading/opencv/modules/videoio/test/test_video_io.cpp:97: progressive seek
/Users/appveyor/projects/opencv-image-reading/opencv/modules/videoio/test/test_video_io.cpp:289: Failure
Expected: (psnr) > (PSNR_GT), actual: 26.8034 vs 30
frame 98
Google Test trace:
/Users/appveyor/projects/opencv-image-reading/opencv/modules/videoio/test/test_video_io.cpp:111: random seek
[  FAILED  ] videoio/videoio_synthetic.write_read_position/39, where GetParam() = (976x768, FOURCC(MJPG), .avi, CV_MJPEG, 30dB) (698 ms)
```

There is a psnr issue. I am trying to solve these. The first one might be related with that i am not able to pick correct simd files for architecture. But i dont think second one is related with that.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt (hunk 4 above)
Comment (alalek) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt]: If this code is borrowed from upstream, then we should keep it with minimal changes (to keep updates from upstream reliable)

Comment (ocpalo) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt]: I borrowed from upstream. It makes sense to keep it as original. What do you mean by minimal changes?

**Comment by ocpalo** [comment]
```cpp
[ RUN      ] videoio/videoio_bunny.read_position/31, where GetParam() = ("mjpg.avi", CV_MJPEG)
Command exited with code -1073740940
```

This does not happen in the linux environment. I cant figure out why this happens in windows

**Comment by vpisarev** [comment]
@ocpalo, we really need to finalize this PR by the end of GSoC. Please, be more active

File: CMakeLists.txt (hunk 1 above)
Comment (ocpalo) [reply CMakeLists.txt:140]: I realized  i do not need that part. I removed but did not pushed yet. It does not affect anything anyway.

File: CMakeLists.txt [hunk 12]
```diff
@@ -221,7 +230,7 @@ OCV_OPTION(BUILD_ZLIB               "Build zlib from source"             (WIN32
 OCV_OPTION(BUILD_TIFF               "Build libtiff from source"          (WIN32 OR ANDROID OR APPLE OR OPENCV_FORCE_3RDPARTY_BUILD) )
 OCV_OPTION(BUILD_OPENJPEG           "Build OpenJPEG from source"         (WIN32 OR ANDROID OR APPLE OR OPENCV_FORCE_3RDPARTY_BUILD) )
 OCV_OPTION(BUILD_JASPER             "Build libjasper from source"        (WIN32 OR ANDROID OR APPLE OR OPENCV_FORCE_3RDPARTY_BUILD) )
-OCV_OPTION(BUILD_JPEG               "Build libjpeg from source"          (WIN32 OR ANDROID OR APPLE OR OPENCV_FORCE_3RDPARTY_BUILD) )
+OCV_OPTION(BUILD_JPEG               "Build libjpeg from source"          (UNIX OR WIN32 OR ANDROID OR APPLE OR OPENCV_FORCE_3RDPARTY_BUILD) )
```
Comment (ocpalo) [inline CMakeLists.txt:233]: Changed this line temporary to build libjpeg-turbo from source in linux too.

**Comment by ocpalo** [comment]
Update: I reverted WITH_LIBJPEG_TURBO_SIMD flag to WITH_SIMD to keep align with the upstream and figure out a proper way to handle libjpeg_turbo internal WITH_SIMD define. It will probably be enough to keep boolean_number macro but anyway. I am trying to minimizes my time lost during the compile times. Once i figure out, it will be straightforward to update the flag. 

iirc, most of the workflow does not have NASM support. Right now, the tests should have pass because the cmake supposed to pick nonsimd source files. I have verified on my machine but it would be good to verify it on workflow too! @asmorkalov @vpisarev , could you please run the workflow?

**Comment by ocpalo** [comment]
Geometric mean (ms)

|Name of Test|withoutsimd.txt|simd.txt|simd.txt vs withoutsimd.txt (x-factor)|
|---|:-:|:-:|:-:|
|Decode::JPEG|67.759|39.150|1.73|
|Encode::JPEG|119.672|33.812|3.54|

**Comment by ocpalo** [comment]
in the workflow

* macOS ARM64
* Ubuntu ARM64
* iOS 
* Android

have assembly compiler and tests are passed.

**Comment by ocpalo** [comment]
@asmorkalov Hi,

Vadim informed me in the last meeting about CPU type and optimization baseline. I have a few comments about this. Let me know what you think so I can do the changes according to your comments. In the `3rdparty/libjpeg-turbo/CMakeLists.txt` you have commented before that i need to reuse OpenCV variables as x86_64,x86 etc. Then alalek wrote a comment that about we should do minimal changes if code is borrowed from upstream(which is true). Should I still use the openCV variables or do not use to keep align with the upstream. 

My second question is, as you said before some CPU's does not support SIMD instructions. Since we provide this SIMD extension as optional(enabled by default), the user can disable this extension. Should I worry about this? I think it will make CMakeLists complicated.

**Comment by asmorkalov** [comment]
Hello @ocpalo Thanks for the great job! I reviewed CMakeLists in libjpeg-turbo repos and tested previous version of the patch. Findings:
- Libjpeg-turbo cmake checks NEON/altivec/DSPr2 support with C compiler checks in CMake. So we are in save, if the code is build, for example, for amrv6. NEON part is not included and CMake reports warning about this. I tested it with old Raspberry Pi and it passes tests correctly. For the builds where NEON support is enabled on compiler side with `-mfpu=neon` NEON is enabled with libjpeg turbo too. I tested it with my Jetson TK1 board (armv7 with NEON). So everything works well with C compiler indirectly. I do not have Power and MIPS boards to check, but the CMake approach is similar and I do not see any reasons for suspicious.
- Libjpeg-turbo always has AVX2 assembly included into build for Intel systems, but the tests do not fail on old system without AVX2 support. Most probably it's dispatched in runtime.

So What I propose to do:
- Presume original CMake as much as possible. Do not wary about duplicate system checks and other specifics.
- Make `WITH_SIMD` cache variable with value `(NOT CV_DISABLE_OPTIMIZATION)` to have precise control for testing.
- Allow libjpeg-turbo for other Unixes like BSD, Solaris, SunOS, etc. It's already supported by original CMake and I do not see any reason to drop them, if all tools hve been found.

**Comment by ocpalo** [comment]
@asmorkalov Thanks for the feedback! 

* I have created a another flag to enable SIMD which is `WITH_LIBJPEG_TURBO_SIMD` and it is value is `NOT CV_DISABLE_OPTIMIZATION`. WITH_SIMD flag is depends on that flag. I think this satisfy the second item in your propose list.
* I did not drop other Unixes support on purpose. If you are talking about [this line](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/CMakeLists.txt#L482), it is related with TurboJPEG api and in OpenCV it is not enabled anyway. Other than that, I updated the compiler flags to enable maximum optimization in SunOS|Pro and clang.

I am trying to solve how to catch if old systems does not support AVX2 assembly support at compile time.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 13]
```diff
@@ -99,20 +143,81 @@ if(WITH_ARITH_DEC)
   set(JPEG_SOURCES ${JPEG_SOURCES} jdarith.c)
 endif()
 
-# No SIMD
-set(JPEG_SOURCES ${JPEG_SOURCES} jsimd_none.c)
+if(MSVC)
```
Comment (alalek) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:146]: > MSVC
> MSVC_IDE

Looks like there is some mess in using of these variables.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 14]
```diff
@@ -27,7 +71,7 @@ check_type_size("unsigned long" SIZEOF_UNSIGNED_LONG)
 
 if(SIZEOF_SIZE_T EQUAL SIZEOF_UNSIGNED_LONG)
   check_c_source_compiles("int main(int argc, char **argv) { unsigned long a = argc;  return __builtin_ctzl(a); }"
-    HAVE_BUILTIN_CTZL)
+          HAVE_BUILTIN_CTZL)
```
Comment (alalek) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:74]: unnecessary whitespace change

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 15]
```diff
@@ -36,7 +80,7 @@ endif()
 if(UNIX)
   # Check for headers
   check_include_files(locale.h HAVE_LOCALE_H)
-  check_include_files(stddef.h HAVE_STDDEF_H)
+  check_include_files(stddef.h HAVE_STDDEF_H)
```
Comment (alalek) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:83]: unnecessary whitespace change

File: CMakeLists.txt [hunk 16]
```diff
@@ -309,6 +309,8 @@ OCV_OPTION(WITH_OPENJPEG "Include JPEG2K support (OpenJPEG)" ON
 OCV_OPTION(WITH_JPEG "Include JPEG support" ON
   VISIBLE_IF TRUE
   VERIFY HAVE_JPEG)
+OCV_OPTION(WITH_LIBJPEG_TURBO_SIMD "Include SIMD extensions for libjpeg-turbo, if available for this platform" (NOT CV_DISABLE_OPTIMIZATION)
+  VISIBLE_IF BUILD_JPEG)
```
Comment (alalek) [inline CMakeLists.txt:313]: Please move this flag into `3rdparty/libjpeg-turbo`

`WITH_` prefix should be replaced to `ENABLE_`
(`WITH_` is used in OpenCV for "auto-detection" of external components/dependencies/libraries)

File: CMakeLists.txt [hunk 17]
```diff
@@ -1341,6 +1343,11 @@ if(WITH_JPEG OR HAVE_JPEG)
     status("    JPEG:" NO)
   elseif(BUILD_JPEG)
     status("    JPEG:" "build-${JPEG_LIBRARY} (ver ${JPEG_LIB_VERSION})")
+    if(WITH_LIBJPEG_TURBO_SIMD)
+      status("        SIMD Support:" "YES")
+    else()
+      status("        SIMD Support:" "NO")
+    endif()
```
Comment (alalek) [inline CMakeLists.txt:1350]: We should distinguish a request from the user and the actual "final" value (which depends on the platform (e.g. MSA is not handled yet), compiler, etc)

Also there is still "libjpeg" (without "-turbo")

File: modules/imgcodecs/perf/perf_jpeg.cpp [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+#include "perf_precomp.hpp"
+
+namespace opencv_test
+{
+    using namespace perf;
+
+    typedef perf::TestBaseWithParam<std::string> JPEG;
+
+    PERF_TEST(JPEG, Decode)
+    {
```
Comment (alalek) [inline modules/imgcodecs/perf/perf_jpeg.cpp:13]: Please avoid indentation in namespaces.

File: modules/imgcodecs/perf/perf_jpeg.cpp [hunk 19]
```diff
@@ -0,0 +1,40 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+#include "perf_precomp.hpp"
+
+namespace opencv_test
+{
+    using namespace perf;
+
+    typedef perf::TestBaseWithParam<std::string> JPEG;
```
Comment (alalek) [inline modules/imgcodecs/perf/perf_jpeg.cpp:10]: > std::string

What is that?

File: 3rdparty/libjpeg-turbo/CMakeLists.txt (hunk 13 above)
Comment (ocpalo) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt:149]: I borrowed these lines from the upstream. Same as `MSVC_IDE` parts.

File: CMakeLists.txt (hunk 17 above)
Comment (ocpalo) [reply CMakeLists.txt]: "libjpeg" without  -turbo will report `"SIMD Support: "NO"` by default. And i think this is not a false promise. I don`t think that i should worry about that. And libjpeg is deprecated according to [docs](https://github.com/opencv/opencv/tree/4.x/3rdparty) at 4.x.

**Comment by ocpalo** [comment]
CMake emits warning when libjpeg-turbo SIMD Extension is enabled but could not activated due to no support. Hence this seems like workflow fails. Should I change it `WARNING` to the `STATUS`?

Also, I was planning to add the following lines to the CMake to see if SIMD extension is supported on compile time. But could not be sure if this is right or not. 

```CMake
if(X86_64)
  if(NOT "${CPU_BASELINE_FINAL}" MATCHES "SSE2|AVX2" OR NOT "${CPU_DISPATCH}" MATCHES "SSE2|AVX2")
    set(WITH_SIMD 0)
  endif()
endif()
if(X86)
  if(NOT "${CPU_BASELINE_FINAL}" MATCHES "MMX|AVX2|SSE2" OR NOT "${CPU_DISPATCH}" MATCHES "MMX|AVX2|SSE2")
    set(WITH_SIMD 0)
  endif()
endif()
```

If someone confirm, i will push it.

**Comment by alalek** [comment]
Only "CPU_BASELINE_FINAL" should be used. Also take a look on other 3rdparty file hows to do this right.

`CPU_DISPATCH` is for dispatching in "runtime" - probably can't be used here due to missing checks (and missing compiler flags).

I believe `NOT CV_DISABLE_OPTIMIZATION` is enough as suggested by @asmorkalov .

**Comment by ocpalo** [comment]
If `NOT CV_DISABLE_OPTIMIZATION` is enough, then I think pr is ready.

```cpp
ENABLE_LIBJPEG_TURBO_SIMD "Include SIMD extensions for libjpeg-turbo, if available for this platform" (NOT CV_DISABLE_OPTIMIZATION)
```

SIMD extension is controlled via ENABLE_LIBJPEG_TURBO_SIMD flag. And this flags value is `NOT CV_DISABLE_OPTIMIZATION`. If there are no magic behind this value, we are good.

cc @alalek @asmorkalov

File: CMakeLists.txt [hunk 20]
```diff
@@ -1341,6 +1341,16 @@ if(WITH_JPEG OR HAVE_JPEG)
     status("    JPEG:" NO)
   elseif(BUILD_JPEG)
     status("    JPEG:" "build-${JPEG_LIBRARY} (ver ${JPEG_LIB_VERSION})")
+    if(ENABLE_LIBJPEG_TURBO_SIMD)
+      status("        SIMD Support Request:" "YES")
+      if(IS_LIBJPEG_TURBO_SIMD_AVAILABLE)
```
Comment (asmorkalov) [inline CMakeLists.txt:1346]: `HAVE_LIBJPEG_TURBO_SIMD` sounds more consistent

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 21]
```diff
@@ -99,12 +146,73 @@ if(WITH_ARITH_DEC)
...
+        string(REGEX REPLACE "-xO3" "-xO5" ${var} "${${var}}")
+      endif()
+      if(${var} MATCHES "-xO2")
+        string(REGEX REPLACE "-xO2" "-xO5" ${var} "${${var}}")
+      endif()
+    endforeach()
+  endif()
+endif()
+
+message(STATUS ${WITH_LIBJPEG_TURBO_SIMD})
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:188]: Please remove debug print

File: 3rdparty/libjpeg-turbo/CMakeLists.txt [hunk 22]
```diff
@@ -99,12 +146,73 @@ if(WITH_ARITH_DEC)
...
+  add_subdirectory(src/simd)
+  if(NEON_INTRINSICS)
+    add_definitions(-DNEON_INTRINSICS)
+  endif()
+else()
+  set(WITH_SIMD 0)
+endif()
+
+if(WITH_SIMD)
+  message(STATUS "SIMD extensions: ${CPU_TYPE} (WITH_SIMD = ${WITH_SIMD})")
```
Comment (asmorkalov) [inline 3rdparty/libjpeg-turbo/CMakeLists.txt:199]: I think this message is redundant. SIMD status is reported form `simd/CmakeLists.txt`. In particular on armv6 I see:
```
SIMD extensions: arm (WITH_SIMD = 1)
```

**Review by asmorkalov — COMMENTED** [review]
Looks good to me except debug messages in CMake.

File: 3rdparty/libjpeg-turbo/CMakeLists.txt (hunk 22 above)
Comment (ocpalo) [reply 3rdparty/libjpeg-turbo/CMakeLists.txt:199]: Actually this line reports this. There is no message reported on `simd/CMakeLists.txt` except failures and it reports NEON or partial NEON support. I believe it is good have for other type CPU's.

**Review by asmorkalov — APPROVED** [review]
:+1: Well Done!

**Comment by ocpalo** [comment]
Thanks, I appreciate for the support and the help!
